### PR TITLE
Fixed bug about boot_order_check when run it with q35

### PR DESCRIPTION
**Pull request by leidwang**
Wait a time interval until get the device pci address

It needs more time to get the complete 'info pci' information, so wait a time interval to get the device pci address.

ID:1820473

Signed-off-by: Leidong Wang <leidwang@redhat.com>

### CODE REVIEW

**Comment by leidwang** [comment]
(1/3) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0.q35: PASS (231.56 s)
 (2/3) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1.q35: PASS (231.53 s)
 (3/3) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2.q35: PASS (236.12 s)

**Comment by leidwang** [comment]
@ybduan Hi,please help me review this patch,thanks!

**Comment by leidwang** [comment]
(01/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0: PASS (229.51 s)
 (02/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0.q35: PASS (233.09 s)
 (03/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1: PASS (229.45 s)
 (04/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1.q35: PASS (232.94 s)
 (05/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2: PASS (233.04 s)
 (06/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2.q35: PASS (236.50 s)
 (07/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0: PASS (229.56 s)
 (08/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0.q35: PASS (231.80 s)
 (09/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1: PASS (229.50 s)
 (10/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1.q35: PASS (231.78 s)
 (11/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2: PASS (233.71 s)
 (12/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2.q35: PASS (235.81 s)

File: qemu/tests/cfg/boot_order_check.cfg [hunk 1]
```diff
@@ -6,6 +6,7 @@
     kill_vm = yes
     boot_menu = on
     enable_sga = yes
+    wait_for_devices_load = 2
```
Comment (ybduan) [inline qemu/tests/cfg/boot_order_check.cfg:9]: It's better to change this variable name to 'devices_load_timeout'.

Comment (leidwang) [reply qemu/tests/cfg/boot_order_check.cfg]: No problem,have done

File: qemu/tests/boot_order_check.py [hunk 2]
```diff
@@ -71,6 +71,9 @@ def _get_pci_addr_by_devid(vm, dev_id):
             process.system("ifconfig %s down" % nic.ifname)
 
     vm.resume()
+    sleep_time = int(params.get("devices_load_timeout"))
+    time.sleep(sleep_time)
+    vm.verify_alive()
```
Comment (ybduan) [inline qemu/tests/boot_order_check.py:77]: The root cause is that we need more time to get the complete pci info. Sleeping here may not resolve this problem.
We can set a acceptable timeout value like 10s to wait for complete pci info:
```Python
def _get_device(devices, dev_id):
    device_found = {}
    for dev in devices:
        if dev['qdev_id'] == dev_id:
            device_found = dev
            break
        elif dev['class_info']['desc'] == 'PCI bridge':
            pci_bridge_devices = dev['pci_bridge'].get('devices')
            if not pci_bridge_devices:
                continue
            device_found = _get_device(pci_bridge_devices,
                                       dev_id)
            if device_found:
                break
   return device_found

def _get_pci_addr_by_devid(dev_id):
    def wrap_func():
        dev_addr = ''
        dev_addr_fmt = '%02d:%02d.%d'
        pci_info = vm.monitor.info('pci', debug=False)
        if isinstance(pci_info, list):
            device = _get_device(pci_info[0]['devices'], dev_id)
            if device:
                dev_addr = dev_addr_fmt % (device['bus'],
                                           device['slot'],
                                           device['function'])
        else:
            # As device id in the last line of info pci output
            # We need reverse the pci information to get the pci addr which is in the
            # front row.
            pci_list = str(pci_info).split("\n")
            pci_list.reverse()
            pci_info = " ".join(pci_list)
            dev_match = re.search(nic_addr_filter % dev_id, pci_info)
            if dev_match:
                bus_slot_func = [int(i) for i in dev_match.groups()]
                dev_addr = dev_addr_fmt % tuple(bus_slot_func)
        return dev_addr
    return wrap_func

pci_addr = utils_misc.wait_for(_get_pci_addr_by_devid(nic.device_id),
                                                devices_load_timeout)
```

**Comment by ybduan** [comment]
Please fix the failing check.

File: qemu/tests/boot_order_check.py [hunk 3]
```diff
@@ -71,6 +76,7 @@ def _get_pci_addr_by_devid(vm, dev_id):
             process.system("ifconfig %s down" % nic.ifname)
 
     vm.resume()
+    devices_load_timeout = int(params.get("devices_load_timeout"))
```
Comment (ybduan) [inline qemu/tests/boot_order_check.py:79]: devices_load_timeout = params.get("devices_load_timeout", 10)

**Comment by ybduan** [comment]
Please update with the latest test results, thanks!

**Comment by leidwang** [comment]
> Please update with the latest test results, thanks!

 (01/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0: PASS (229.68 s)
 (02/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0.q35: PASS (233.47 s)
 (03/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1: PASS (229.83 s)
 (04/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1.q35: PASS (233.30 s)
 (05/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2: PASS (233.46 s)
 (06/12) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2.q35: PASS (236.87 s)
 (07/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0: PASS (229.90 s)
 (08/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0.q35: PASS (232.32 s)
 (09/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1: PASS (229.96 s)
 (10/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1.q35: PASS (232.41 s)
 (11/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2: PASS (233.10 s)
 (12/12) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2.q35: PASS (235.29 s)

**Comment by ybduan** [comment]
Please help review this PR, thanks! @XueqiangWei

File: qemu/tests/boot_order_check.py [hunk 4]
```diff
@@ -71,6 +76,7 @@ def _get_pci_addr_by_devid(vm, dev_id):
             process.system("ifconfig %s down" % nic.ifname)
 
     vm.resume()
+    devices_load_timeout = params.get("devices_load_timeout", 10)
```
Comment (XueqiangWei) [inline qemu/tests/boot_order_check.py:79]: should convert it to integer by int()

Comment (ybduan) [reply qemu/tests/boot_order_check.py:79]: It will be converted to float in utils_misc.wait_for() automatically.

Comment (XueqiangWei) [reply qemu/tests/boot_order_check.py]: I see, I think it's more graceful when the default value is integer and the get value is string.

**Comment by ybduan** [comment]
Please update the commit message to be more specific. Like:
"Wait a time interval until get the device pci address
It needs more time to get the complete 'info pci' information, so wait a time interval to get the device pci address."

File: qemu/tests/boot_order_check.py [hunk 5]
```diff
@@ -29,34 +29,39 @@ def _get_device(devices, dev_id):
...
-            pci_list = str(pci_info).split("\n")
-            pci_list.reverse()
-            pci_info = " ".join(pci_list)
-            dev_match = re.search(nic_addr_filter % dev_id, pci_info)
-            if dev_match:
-                bus_slot_func = [int(i) for i in dev_match.groups()]
-                dev_addr = dev_addr_fmt % tuple(bus_slot_func)
-        return dev_addr
+    def _get_pci_addr_by_devid(dev_id):
+        def wrap_func():
```
Comment (XueqiangWei) [inline qemu/tests/boot_order_check.py:42]: I think no need \"wrap_func()\".  just use it like 
pci_addr = utils_misc.wait_for(lambda: _get_pci_addr_by_devid(nic.device_id),
                                                 timeout=devices_load_timeout)

File: qemu/tests/boot_order_check.py [hunk 6]
```diff
@@ -83,7 +89,8 @@ def _get_pci_addr_by_devid(vm, dev_id):
 
     for nic in vm.virtnet:
         boot_index = params['bootindex_%s' % nic.nic_name]
-        pci_addr = _get_pci_addr_by_devid(vm, nic.device_id)
+        pci_addr = utils_misc.wait_for(_get_pci_addr_by_devid(nic.device_id),
+                                       devices_load_timeout)
```
Comment (XueqiangWei) [inline qemu/tests/boot_order_check.py:93]: pci_addr = utils_misc.wait_for(lambda: _get_pci_addr_by_devid(nic.device_id), timeout=devices_load_timeout)

File: qemu/tests/boot_order_check.py [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 from avocado.utils import process
 
-from virttest import error_context
+from virttest import error_context, utils_misc
```
Comment (XueqiangWei) [inline qemu/tests/boot_order_check.py:7]: from virttest import utils_misc

**Comment by XueqiangWei** [comment]
@leidwang four commits merge  into one.

**Comment by XueqiangWei** [comment]
(1/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0.q35: PASS (244.53 s)
 (2/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1.q35: PASS (243.75 s)
 (3/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2.q35: PASS (249.60 s)
 (4/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder0.q35: PASS (244.05 s)
 (5/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder1.q35: PASS (242.40 s)
 (6/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.boot_order_check.bootorder2.q35: PASS (248.02 s)

Acked-by: xuwei@redhat.com

**Comment by XueqiangWei** [comment]
@luckyh please help review it. Many thanks.

**Comment by ybduan** [comment]
Acked-by: Yanbin Duan yduan@redhat.com

**Comment by luckyh** [comment]
LGTM.

Btw,  I see there is the note `root committed 18 days ago` in the commit tab view, so it seems that your git configuration wasn't set up properly, please follow the below references and commit the change again, thanks!

https://help.github.com/en/github/using-git/setting-your-username-in-git#setting-your-git-username-for-a-single-repository
https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address#setting-your-commit-email-address-in-git

**Comment by leidwang** [comment]
> LGTM.
> 
> Btw, I see there is the note `root committed 18 days ago` in the commit tab view, so it seems that your git configuration wasn't set up properly, please follow the below references and commit the change again, thanks!
> 
> https://help.github.com/en/github/using-git/setting-your-username-in-git#setting-your-git-username-for-a-single-repository
> https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address#setting-your-commit-email-address-in-git

Thanks for your reminder, Has changed.

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**